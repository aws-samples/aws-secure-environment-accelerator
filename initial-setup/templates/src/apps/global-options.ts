import * as cdk from '@aws-cdk/core';
import { NonEmptyString } from 'io-ts-types/lib/NonEmptyString';
import { AcceleratorNameTagger } from '@aws-pbmm/common-cdk/lib/core/name-tagger';
import { GlobalOptions } from '../global-options/stack';
import { getAccountId, loadAccounts } from '../utils/accounts';
import { loadAcceleratorConfig } from '../utils/config';
import { loadContext } from '../utils/context';
import { loadStackOutputs } from '../utils/outputs';

process.on('unhandledRejection', (reason, _) => {
  console.error(reason);
  process.exit(1);
});

async function main() {
  const context = loadContext();
  const acceleratorConfig = await loadAcceleratorConfig();
  const accounts = await loadAccounts();
  const outputs = await loadStackOutputs();

  const globalOptionsConfig = acceleratorConfig['global-options'];

  // TODO Get these values dynamically
  const zonesConfig = globalOptionsConfig.zones;
  const [zoneAccount, resolverVpc, resolverSubnet] = [
    zonesConfig.account,
    zonesConfig['resolver-vpc'],
    zonesConfig['resolver-subnet'],
  ];
  const zonesCreationAccountId = getAccountId(accounts, zoneAccount);

  const app = new cdk.App();

  new GlobalOptions.Stack(app, 'GlobalOptions', {
    env: {
      account: zonesCreationAccountId,
      region: cdk.Aws.REGION,
    },
    acceleratorName: context.acceleratorName,
    acceleratorPrefix: context.acceleratorPrefix,
    stackName: 'PBMMAccel-GlobalOptions',
    context,
    acceleratorConfig,
    outputs,
  });

  // Add accelerator tag to all resources
  cdk.Tag.add(app, 'Accelerator', context.acceleratorName);

  // Add name tag to all resources
  app.node.applyAspect(new AcceleratorNameTagger());
}

// tslint:disable-next-line: no-floating-promises
main();
