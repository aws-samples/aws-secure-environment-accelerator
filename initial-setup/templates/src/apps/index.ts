export * from './shared';
export * from './phase-0';
export * from './phase-1';
export * from './phase-2';
export * from './phase-3';
export * from './phase-4';
export * from './phase-5';
