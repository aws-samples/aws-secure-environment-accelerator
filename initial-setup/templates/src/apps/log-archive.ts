import * as cdk from '@aws-cdk/core';
import { OrganizationalUnit } from '../organizational-units/stack';
import { LogArchive } from '../log-archive/stack';
import { getAccountId, loadAccounts } from '../utils/accounts';
import { loadAcceleratorConfig } from '../utils/config';
import { loadContext } from '../utils/context';

process.on('unhandledRejection', (reason, _) => {
  console.error(reason);
  process.exit(1);
});

async function main() {
  const context = loadContext();
  const acceleratorConfig = await loadAcceleratorConfig();
  const accounts = await loadAccounts();

  // TODO Get these values dynamically
  const globalOptionsConfig = acceleratorConfig['global-options'];
  const centralLogRetention = globalOptionsConfig['central-log-retention'];
  const logArchiveAccountId = getAccountId(accounts, 'log-archive');
  const sharedNetworkAccountId = getAccountId(accounts, 'shared-network');

  const app = new cdk.App();

  const logArchiveStack = new LogArchive.Stack(app, 'LogArchive', {
    env: {
      account: logArchiveAccountId,
      region: cdk.Aws.REGION,
    },
    acceleratorName: context.acceleratorName,
    acceleratorPrefix: context.acceleratorPrefix,
    stackName: 'PBMMAccel-LogArchive',
    centralLogRetentionInDays: centralLogRetention,
    sharedNetWorkAccountId: sharedNetworkAccountId,
  });

  // store the s3 bucket arn for later reference
  new cdk.CfnOutput(logArchiveStack, 's3BucketArn', {
    value: logArchiveStack.s3BucketArn,
  });

  // store the s3 bucket - kms key arn for later reference
  new cdk.CfnOutput(logArchiveStack, 's3KmsKeyArn', {
    value: logArchiveStack.s3KmsKeyArn,
  });
}

// tslint:disable-next-line: no-floating-promises
main();
