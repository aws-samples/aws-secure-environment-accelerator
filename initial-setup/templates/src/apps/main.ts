import * as cdk from '@aws-cdk/core';
import { AcceleratorStack, AcceleratorStackProps } from '@aws-pbmm/common-cdk/lib/core/accelerator-stack';
import { AccountConfig, OrganizationalUnitConfig, VpcConfig } from '@aws-pbmm/common-lambda/lib/config';
import { pascalCase } from 'pascal-case';
import { Context } from 'vm';
import { Vpc } from '../common/vpc';
import { Account, getAccountId, loadAccounts } from '../utils/accounts';
import { loadAcceleratorConfig } from '../utils/config';
import { loadContext } from '../utils/context';
import { throws } from 'assert';
import { FlowLogBucket } from '../common/flow-log-bucket';

process.on('unhandledRejection', (reason, _) => {
  console.error(reason);
  process.exit(1);
});

async function main() {
  const context = loadContext();
  const acceleratorConfig = await loadAcceleratorConfig();
  const accounts = await loadAccounts();

  // const logArchiveAccountId = getAccountId(accounts, 'log-archive');
  // const logArchiveS3BucketArn = getStackOutput(outputs, 'log-archive', 's3BucketArn');
  // const logArchiveS3KmsKeyArn = getStackOutput(outputs, 'log-archive', 's3KmsKeyArn');

  const app = new cdk.App();

  // const accountStacks: { [accountKey: string]: AcceleratorStack } = {};
  // const getVpcStackForAccount = (accountKey: string): AcceleratorStack => {
  //   if (accountStacks[accountKey]) {
  //     return accountStacks[accountKey];
  //   }

  //   const name = pascalCase(accountKey);
  //   const stack = new AcceleratorStack(app, `Networking${name}`, {
  //     env: {
  //       account: getAccountId(accounts, accountKey),
  //       region: cdk.Aws.REGION,
  //     },
  //     stackName: `PBMMAccel-Networking${name}`,
  //     acceleratorName: context.acceleratorName,
  //     acceleratorPrefix: context.acceleratorPrefix,
  //   });

  //   accountStacks[accountKey] = stack;
  //   return stack;
  // };

  // Create all the VPCs for the mandatory accounts
  const mandatoryAccountDeployments: { [accountKey: string]: MandatoryAccountDeployment } = {};
  const mandatoryAccountConfig = acceleratorConfig['mandatory-account-configs'];
  for (const [accountKey, accountConfig] of Object.entries(mandatoryAccountConfig)) {
    const name = pascalCase(accountKey);
    const deployment = new MandatoryAccountDeployment(app, name, {
      context,
      accounts,
      accountKey,
      accountConfig,
    });
    mandatoryAccountDeployments[accountKey] = deployment;
  }

  // Create all the VPCs for the organizational units
  const organizationalUnits = acceleratorConfig['organizational-units'];
  for (const [ouKey, ouConfig] of Object.entries(organizationalUnits)) {
    new OrganizationalUnitDeployment(app, ouKey, {
      context,
      accounts,
      ouKey,
      ouConfig,
      mandatoryAccountDeployments,
    });
  }
}

// tslint:disable-next-line: no-floating-promises
main();

export type VpcStackProps = AcceleratorStackProps;

export class VpcStack extends AcceleratorStack {
  flowLogBucket: FlowLogBucket | undefined;

  constructor(scope: cdk.Construct, id: string, props: VpcStackProps) {
    super(scope, id, props);
  }

  getOrCreateFlowLogBucket(): FlowLogBucket {
    if (!this.flowLogBucket) {
      this.flowLogBucket = new FlowLogBucket(this, 'FlowLogBucket', {
        // TODO From props
        expirationInDays: 30,
      });
    }
    return this.flowLogBucket;
  }
}

interface MandatoryAccountDeploymentProps {
  context: Context;
  accounts: Account[];
  accountKey: string;
  accountConfig: AccountConfig;
}

class MandatoryAccountDeployment extends cdk.Construct {
  readonly accountConfig: AccountConfig;
  readonly vpcStack: VpcStack;

  constructor(scope: cdk.Construct, id: string, props: MandatoryAccountDeploymentProps) {
    super(scope, id);

    const { context, accounts, accountKey } = props;

    this.accountConfig = props.accountConfig;

    this.vpcStack = new VpcStack(this, `VpcStack`, {
      env: {
        account: getAccountId(accounts, accountKey),
        region: cdk.Aws.REGION,
      },
      stackName: `PBMMAccel-Networking${pascalCase(accountKey)}`,
      acceleratorName: context.acceleratorName,
      acceleratorPrefix: context.acceleratorPrefix,
    });

    const vpcConfig: VpcConfig | undefined = this.accountConfig.vpc;
    if (!vpcConfig) {
      console.log(`Skipping VPC creation for account "${accountKey}"`);
    } else if (vpcConfig.deploy !== 'local') {
      console.warn(`Skipping non-local VPC deployment for mandatory account "${accountKey}"`);
    } else {
      console.debug(`Deploying VPC in account "${accountKey}"`);

      new Vpc(this.vpcStack, vpcConfig.name, {
        accounts,
        vpcConfig,
      });
    }
  }
}

interface OrganizationalUnitDeploymentProps {
  context: Context;
  accounts: Account[];
  ouKey: string;
  ouConfig: OrganizationalUnitConfig;
  mandatoryAccountDeployments: { [accountKey: string]: MandatoryAccountDeployment };
}

class OrganizationalUnitDeployment extends cdk.Construct {
  constructor(scope: cdk.Construct, id: string, props: OrganizationalUnitDeploymentProps) {
    super(scope, id);

    const { accounts, ouKey, ouConfig, mandatoryAccountDeployments } = props;

    const vpcConfig = ouConfig?.vpc;
    const deploy = vpcConfig?.deploy;
    if (!vpcConfig) {
      console.log(`Skipping VPC creation for organizational unit "${ouKey}"`);
    } else if (!deploy) {
      console.warn(`Skipping VPC creation for organizational unit "${ouKey}" as 'deploy' is not set`);
    } else if (deploy === 'local') {
      // If the deployment is 'local' then the VPC should be created in all the accounts in this OU
      for (const [accountKey, accountDeployment] of Object.entries(mandatoryAccountDeployments)) {
        if (accountDeployment.accountConfig.ou !== ouKey) {
          // Skip OU VPC deployment if the account is not in the OU
          continue;
        }

        console.debug(`Deploying local VPC for organizational unit "${ouKey}" in account "${accountKey}"`);

        new Vpc(accountDeployment.vpcStack, vpcConfig.name, {
          accounts,
          vpcConfig,
          organizationalUnitName: ouKey,
        });
      }
    } else {
      // If the deployment is not 'local' then the VPC should be created in the given account
      const accountKey = deploy;
      console.debug(`Deploying non-local VPC for organizational unit "${ouKey}" in account "${accountKey}"`);

      const accountDeployment = mandatoryAccountDeployments[accountKey];
      new Vpc(accountDeployment.vpcStack, vpcConfig.name, {
        accounts,
        vpcConfig,
        organizationalUnitName: ouKey,
      });
    }
  }
}
