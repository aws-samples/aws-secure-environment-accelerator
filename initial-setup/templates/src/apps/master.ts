import * as cdk from '@aws-cdk/core';
import { AcceleratorNameTagger } from '@aws-pbmm/common-cdk/lib/core/name-tagger';
import { Master } from '../master/stack';
import { getAccountId, loadAccounts } from '../utils/accounts';
import { loadAcceleratorConfig } from '../utils/config';
import { loadContext } from '../utils/context';
import { getStackOutput, loadStackOutputs } from '../utils/outputs';

process.on('unhandledRejection', (reason, _) => {
  console.error(reason);
  process.exit(1);
});

async function main() {
  const context = loadContext();
  const acceleratorConfig = await loadAcceleratorConfig();
  const accounts = await loadAccounts();
  const outputs = await loadStackOutputs();

  const mandatoryAccountConfig = acceleratorConfig['mandatory-account-configs'];

  // TODO Get these values dynamically
  const masteAccountAccountId = getAccountId(accounts, 'master');
  const masteAccountConfig = mandatoryAccountConfig['master'];

  const app = new cdk.App();

  new Master.Stack(app, 'Master', {
    env: {
      account: masteAccountAccountId,
      region: cdk.Aws.REGION,
    },
    stackName: 'PBMMAccel-Master',
    accountConfig: masteAccountConfig,
  });

  // Add accelerator tag to all resources
  cdk.Tag.add(app, 'Accelerator', context.acceleratorName);

  // Add name tag to all resources
  app.node.applyAspect(new AcceleratorNameTagger());
}

// tslint:disable-next-line: no-floating-promises
main();
