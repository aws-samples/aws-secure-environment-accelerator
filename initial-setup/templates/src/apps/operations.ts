import * as cdk from '@aws-cdk/core';
import { getAccountId, loadAccounts } from '../utils/accounts';
import { loadAcceleratorConfig } from '../utils/config';
import { Operations } from '../operations/stack';
import { SecretsStack } from '../../../../common-cdk/lib/core/secrets-stack';
import { loadContext } from '../utils/context';
import * as iam from '@aws-cdk/aws-iam';

process.on('unhandledRejection', (reason, _) => {
  console.error(reason);
  process.exit(1);
});

async function main() {
  const context = loadContext();
  const acceleratorConfig = await loadAcceleratorConfig();
  const accounts = await loadAccounts();

  const mandatoryAccountConfig = acceleratorConfig['mandatory-account-configs'];

  // TODO get the subnet ids from respective account
  // const subnetInfo = getStackOutput(outputs, 'shared-network', 'SubnetInfo');
  const subnetInfo = {
    vpcId: 'vpc-0006f352d764d0c78',
    subnetIds: ['subnet-00d10366607bf20b1', 'subnet-0e412cda1d44e8537'],
  };

  const app = new cdk.App();

  for (const [orgKey, accountConfig] of Object.entries(mandatoryAccountConfig)) {
    const madDeploymentConfig = accountConfig.deployments.mad!;
    if (madDeploymentConfig && madDeploymentConfig.deploy) {
      const accountId = getAccountId(accounts, accountConfig['account-name']);

      const secretsStack = new SecretsStack(app, 'SharedNetworkSecrets', {
        env: {
          account: getAccountId(accounts, 'master'),
          region: cdk.Aws.REGION,
        },
        acceleratorName: context.acceleratorName,
        acceleratorPrefix: context.acceleratorPrefix,
        stackName: `PBMMAccel-${accountConfig['account-name']}Secrets`,
      });

      const madPassword = secretsStack.createSecret('MadPassword', {
        secretName: `accelerator/${accountConfig['account-name']}/mad/password`,
        description: 'Password for Managed Active Directory.',
        generateSecretString: {
          passwordLength: 16,
        },
        principals: [new iam.ArnPrincipal(`arn:aws:iam::${accountId}:role/${context.acceleratorExecutionRoleName}`)],
      });

      new Operations.Stack(app, 'Operations', {
        env: {
          account: accountId,
          region: cdk.Aws.REGION,
        },
        stackName: `PBMMAccel-${accountConfig['account-name']}`,
        madDeploymentConfig,
        subnetInfo,
        password: madPassword,
      });
    }
  }
}

// tslint:disable-next-line: no-floating-promises
main();
