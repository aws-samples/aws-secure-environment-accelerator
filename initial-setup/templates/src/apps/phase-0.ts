import * as path from 'path';
import { pascalCase } from 'pascal-case';
import * as cdk from '@aws-cdk/core';
import * as iam from '@aws-cdk/aws-iam';
import * as lambda from '@aws-cdk/aws-lambda';
import * as s3 from '@aws-cdk/aws-s3';
import * as s3deployment from '@aws-cdk/aws-s3-deployment';
import { createBucketName } from '@aws-pbmm/common-cdk/lib/core/accelerator-name-generator';
import * as outputKeys from '@aws-pbmm/common-outputs/lib/stack-output';
import { getAccountId, loadAccounts, Account } from '../utils/accounts';
import { loadAcceleratorConfig } from '../utils/config';
import { loadContext } from '../utils/context';
import { LogArchiveBucket } from '../common/log-archive-bucket';
import { AccountStacks } from '../common/account-stacks';
import { JsonOutputValue } from '../common/json-output';
import { SecurityHubStack } from '../common/security-hub';
import { AccessAnalyzer } from '../common/access-analyzer';
import * as centralServices from '../deployments/central-services';
import * as defaults from '../deployments/defaults';
import * as firewallCluster from '../deployments/firewall/cluster';
import * as mad from '../deployments/mad';
import * as budget from '../deployments/billing/budget';

process.on('unhandledRejection', (reason, _) => {
  console.error(reason);
  process.exit(1);
});

/**
 * This is the main entry point to deploy phase 0.
 *
 * The following resources are deployed in phase 0:
 *   - Log archive bucket
 *
 * TODO This phase could be merged into phase 1.
 */
async function main() {
  const context = loadContext();
  const acceleratorConfig = await loadAcceleratorConfig();
  const accounts = await loadAccounts();

  // TODO Get these values dynamically
  const globalOptionsConfig = acceleratorConfig['global-options'];
  const mandatoryAccountConfig = acceleratorConfig.getMandatoryAccountConfigs();
  const logRetentionInDays = globalOptionsConfig['central-log-retention'];
  // TODO Remove hard-coded 'log-archive' account key and use configuration file somehow
  const logArchiveAccountId = getAccountId(accounts, 'log-archive');

  const app = new cdk.App();

  const accountStacks = new AccountStacks(app, {
    phase: 0,
    accounts,
    context,
  });

  // Master Stack to update Custom Resource Lambda Functions invoke permissions
  // TODO Remove hard-coded 'master' account key and use configuration file somehow
  const masterAccountStack = accountStacks.getOrCreateAccountStack('master');
  for (const [index, funcArn] of Object.entries(context.cfnCustomResourceFunctions)) {
    for (const account of accounts) {
      new lambda.CfnPermission(masterAccountStack, `${index}${account.key}InvokePermission`, {
        functionName: funcArn,
        action: 'lambda:InvokeFunction',
        principal: `arn:aws:iam::${account.id}:role/${context.acceleratorExecutionRoleName}`,
      });
    }
  }

  // TODO Remove hard-coded 'log-archive' account key and use configuration file somehow
  const logArchiveStack = accountStacks.getOrCreateAccountStack('log-archive');

  const accountIds: string[] = accounts.map(account => account.id);

  // Create the log archive bucket
  const bucket = new LogArchiveBucket(logArchiveStack, 'LogArchive', {
    logRetention: cdk.Duration.days(logRetentionInDays),
    logArchiveAccountId,
    accountIds,
  });

  // Grant all accounts access to the log archive bucket
  const principals = accounts.map(account => new iam.AccountPrincipal(account.id));
  bucket.grantReplicate(...principals);

  // store the log archive account Id for later reference
  new cdk.CfnOutput(logArchiveStack, outputKeys.OUTPUT_LOG_ARCHIVE_ACCOUNT_ID, {
    value: logArchiveAccountId,
  });

  // store the s3 bucket arn for later reference
  new cdk.CfnOutput(logArchiveStack, outputKeys.OUTPUT_LOG_ARCHIVE_BUCKET_ARN, {
    value: bucket.bucketArn,
  });

  // store the s3 bucket - kms key arn for later reference
  new cdk.CfnOutput(logArchiveStack, outputKeys.OUTPUT_LOG_ARCHIVE_ENCRYPTION_KEY_ARN, {
    value: bucket.encryptionKeyArn,
  });

  // store the s3 bucket - kms key id for later reference
  new cdk.CfnOutput(logArchiveStack, outputKeys.OUTPUT_LOG_ARCHIVE_ENCRYPTION_KEY_ID, {
    value: bucket.encryptionKey.keyId,
  });

  // TODO Replace above outputs with JSON output
  // new JsonOutputValue(stack, 'LogArchiveOutput', {
  //   type: 'LogArchiveOutput',
  //   value: {
  //     bucketArn: bucket.bucketArn,
  //     encryptionKeyArn: bucket.encryptionKeyArn,
  //   },
  // });

  const uploadArtifacts = (
    artifactName: string,
    artifactFolderName: string,
    artifactKeyPrefix: string,
    accountKey: string,
    bucketNameGeneratorInput: string,
    destinationKeyPrefix?: string,
  ): void => {
    // creating a bucket to store artifacts
    const artifactBucket = new s3.Bucket(masterAccountStack, `${artifactName}ArtifactsBucket${accountKey}`, {
      versioned: true,
      bucketName: createBucketName(bucketNameGeneratorInput),
    });

    // Granting read access to all the accounts
    principals.map(principal => artifactBucket.grantRead(principal));

    const artifactsFolderPath = path.join(__dirname, '..', '..', '..', '..', 'reference-artifacts', artifactFolderName);

    new s3deployment.BucketDeployment(masterAccountStack, `${artifactName}ArtifactsDeployment${accountKey}`, {
      sources: [s3deployment.Source.asset(artifactsFolderPath)],
      destinationBucket: artifactBucket,
      destinationKeyPrefix,
    });

    // outputs to store reference artifacts s3 bucket information
    new JsonOutputValue(masterAccountStack, `${artifactName}ArtifactsOutput${accountKey}`, {
      type: `${artifactName}ArtifactsOutput`,
      value: {
        accountKey,
        bucketArn: artifactBucket.bucketArn,
        bucketName: artifactBucket.bucketName,
        keyPrefix: artifactKeyPrefix,
      },
    });
  };

  // upload IAM-Policies Artifacts
  uploadArtifacts('IamPolicy', 'iam-policies', 'iam-policy', 'master', 'iam', 'iam-policy');

  const createAccessAnalyzer = async (accountKey: string): Promise<void> => {
    const accountStack = accountStacks.getOrCreateAccountStack(accountKey);

    new AccessAnalyzer(accountStack, `Access Analyzer-${pascalCase(accountKey)}`);
  };

  // creating assets for default account settings
  for (const [accountKey, accountConfig] of mandatoryAccountConfig) {
    // TODO Remove hard-coded account key
    if (accountKey === 'security') {
      await createAccessAnalyzer(accountKey);
    }
  }

  for (const [accountKey, accountConfig] of acceleratorConfig.getAccountConfigs()) {
    const madDeploymentConfig = accountConfig.deployments?.mad;
    if (!madDeploymentConfig || !madDeploymentConfig.deploy) {
      continue;
    }
    const mandatoryAccountId = getAccountId(accounts, accountKey);
    const rdgwBucketName = `pbmmaccel-${mandatoryAccountId}-${cdk.Aws.REGION}`;

    // upload RDGW Artifacts
    uploadArtifacts('Rdgw', 'scripts', 'config/scripts/', accountKey, 'rdgw', 'config/scripts');
  }

  const globalOptions = acceleratorConfig['global-options'];
  const securityMasterAccount = accounts.find(a => a.type === 'security' && a.ou === 'core');
  const subAccountIds = accounts.map(account => {
    return {
      AccountId: account.id,
      Email: account.email,
    };
  });
  const securityMasterAccountStack = accountStacks.getOrCreateAccountStack(securityMasterAccount?.key!);
  // Create Security Hub stack for Master Account in Security Account
  const securityHubMaster = new SecurityHubStack(securityMasterAccountStack, `SecurityHubMasterAccountSetup`, {
    account: securityMasterAccount!,
    acceptInvitationFuncArn: context.cfnCustomResourceFunctions.acceptInviteSecurityHubFunctionArn,
    enableStandardsFuncArn: context.cfnCustomResourceFunctions.enableSecurityHubFunctionArn,
    inviteMembersFuncArn: context.cfnCustomResourceFunctions.inviteMembersSecurityHubFunctionArn,
    standards: globalOptions['security-hub-frameworks'],
    subAccountIds,
  });

  // Create defaults, e.g. S3 buckets, EBS encryption keys
  const defaultsResult = await defaults.step1({
    acceleratorName: context.acceleratorName,
    accountStacks,
    config: acceleratorConfig,
  });

  // MAD creation step 1
  // Needs EBS default keys from the EBS default step
  await mad.step1({
    acceleratorName: context.acceleratorName,
    acceleratorPrefix: context.acceleratorPrefix,
    accountEbsEncryptionKeys: defaultsResult.accountEbsEncryptionKeys,
    accountStacks,
    config: acceleratorConfig,
  });

  // Firewall creation step 1
  await firewallCluster.step1({
    accountStacks,
    config: acceleratorConfig,
  });

  // Budget creation step 1
  await budget.step1({
    accountStacks,
    config: acceleratorConfig,
  });

  // Central Services step 1
  await centralServices.step1({
    accountStacks,
    config: acceleratorConfig,
    accounts,
  });
}

// tslint:disable-next-line: no-floating-promises
main();
