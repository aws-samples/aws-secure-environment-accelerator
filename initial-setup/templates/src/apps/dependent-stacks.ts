import * as cdk from '@aws-cdk/core';
import { getAccountId, loadAccounts } from '../utils/accounts';
import { loadAcceleratorConfig } from '../utils/config';
import { loadContext } from '../utils/context';
import { loadStackOutputs } from '../utils/outputs';
import { DependentResources } from '../common/dependent-resources-stack';
import { GlobalOptionsDeployment } from '../common/global-options';

process.on('unhandledRejection', (reason, _) => {
  console.error(reason);
  process.exit(1);
});

async function main() {
  const context = loadContext();
  const acceleratorConfig = await loadAcceleratorConfig();
  const accounts = await loadAccounts();
  const outputs = await loadStackOutputs();

  const app = new cdk.App();

  const globalOptionsConfig = acceleratorConfig['global-options'];
  const zonesConfig = globalOptionsConfig.zones;

  const deployment = new DependentResources(app, 'GlobalOptionsDNSResolversStack', {
    env: {
      account: getAccountId(accounts, zonesConfig.account),
      region: cdk.Aws.REGION,
    },
    stackName: `PBMMAccel-GlobalOptionsDNSResolvers`,
    acceleratorName: context.acceleratorName,
    acceleratorPrefix: context.acceleratorPrefix,
  });

  const globalOptionsStack = new GlobalOptionsDeployment(deployment, `GlobalOptionsDNSResolvers`, {
    accounts,
    outputs,
    context,
    acceleratorConfig,
  });
  for (const [key, value] of globalOptionsStack.outputs) {
    new cdk.CfnOutput(deployment, key, {
      value,
    });
  }
}

// tslint:disable-next-line: no-floating-promises
main();
