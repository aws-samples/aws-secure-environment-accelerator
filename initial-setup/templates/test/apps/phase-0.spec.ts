// tslint:disable:no-any
import 'jest';
import { resourcesToList, stackToCloudFormation } from '../jest';
import * as phase0 from '../../src/apps/phase-0';
import { createPhaseInput } from './phase-input';

test('the bucket names and IDs in phase 0 should not change', async () => {
  const input = createPhaseInput('0');

  await phase0.deploy(input);

  for (const stack of input.accountStacks.stacks) {
    // Convert the stack to a CloudFormation template
    const template = stackToCloudFormation(stack);
    const resources = resourcesToList(template.Resources);

    // Only take a snapshot of buckets and LogicalId and BucketName
    const expected = resources
      .filter(r => r.Type === 'AWS::S3::Bucket')
      .map(b => ({
        LogicalId: b.LogicalId,
        Properties: {
          BucketName: b.Properties.BucketName,
        },
      }));

    expect(expected).toMatchSnapshot(stack.stackName);
  }
});
