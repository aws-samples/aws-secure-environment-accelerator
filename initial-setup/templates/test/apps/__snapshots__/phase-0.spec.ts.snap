// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the bucket names and IDs in phase 0 should not change: PBMMAccel-LogArchive-Phase0-bucket-ids 1`] = `
Array [
  "DefaultBucketEEDD2DB7",
  "AesBucket0FF6308D",
]
`;

exports[`the bucket names and IDs in phase 0 should not change: PBMMAccel-LogArchive-Phase0-bucket-names 1`] = `
Array [
  Object {
    "Fn::Join": Array [
      "",
      Array [
        "pbmmaccel-",
        Object {
          "Ref": "AWS::AccountId",
        },
        "-",
        Object {
          "Ref": "AWS::Region",
        },
        "-c25e0f50",
      ],
    ],
  },
  Object {
    "Fn::Join": Array [
      "",
      Array [
        "pbmmaccel-",
        Object {
          "Ref": "AWS::AccountId",
        },
        "-",
        Object {
          "Ref": "AWS::Region",
        },
        "-aes-c25e0f50",
      ],
    ],
  },
]
`;

exports[`the bucket names and IDs in phase 0 should not change: PBMMAccel-Master-Phase0-bucket-ids 1`] = `
Array [
  "CentralBucketCopyDB3C4877",
]
`;

exports[`the bucket names and IDs in phase 0 should not change: PBMMAccel-Master-Phase0-bucket-names 1`] = `
Array [
  Object {
    "Fn::Join": Array [
      "",
      Array [
        "pbmmaccel-",
        Object {
          "Ref": "AWS::AccountId",
        },
        "-",
        Object {
          "Ref": "AWS::Region",
        },
        "-config-1210e7d1",
      ],
    ],
  },
]
`;

exports[`the bucket names and IDs in phase 0 should not change: PBMMAccel-Operations-Phase0-bucket-ids 1`] = `Array []`;

exports[`the bucket names and IDs in phase 0 should not change: PBMMAccel-Operations-Phase0-bucket-names 1`] = `Array []`;

exports[`the bucket names and IDs in phase 0 should not change: PBMMAccel-Security-Phase0-bucket-ids 1`] = `Array []`;

exports[`the bucket names and IDs in phase 0 should not change: PBMMAccel-Security-Phase0-bucket-names 1`] = `Array []`;

exports[`the bucket names and IDs in phase 0 should not change: PBMMAccel-SharedNetwork-Phase0-bucket-ids 1`] = `Array []`;

exports[`the bucket names and IDs in phase 0 should not change: PBMMAccel-SharedNetwork-Phase0-bucket-ids 2`] = `Array []`;

exports[`the bucket names and IDs in phase 0 should not change: PBMMAccel-SharedNetwork-Phase0-bucket-names 1`] = `Array []`;

exports[`the bucket names and IDs in phase 0 should not change: PBMMAccel-SharedNetwork-Phase0-bucket-names 2`] = `Array []`;
